### PR TITLE
based on machine (pi5)  set pwmchip to 2, Also set provider priority …

### PR DESCRIPTION
**Pull request by taartspi**
…of Gpiod input/output, linuxFS i2c & pwm,  so they are loaded and Pigpio is not.  On all legacy Pi (! Pi5) do the opposite priority and pwmchip is 0. I still have the WEB pages updates to do.            

@eitch Result, I also had to use the static way to set the Board info for selecting the pwmchip value.  Repeated runs showed occasionally the Context was not set in the class, same as I hit with getPriority.                                                                                                                                                                           

I decided on Pi5 to make the GpioD provider a higher priority then the linuxfs digital.   Both higher than Pigpio on Pi5.  Below are the providers loaded when all are available 

```
Pi4
---------------------------------------------------------
PI4J PROVIDERS
----------------------------------------------------------
PROVIDERS: [6] "I/O Providers" <com.pi4j.provider.impl.DefaultProviders> 
├─ANALOG_INPUT: [0] <com.pi4j.io.gpio.analog.AnalogInputProvider> 
├─PWM: [1] <com.pi4j.io.pwm.PwmProvider> 
│ └─PROVIDER: "PiGpio PWM Provider" {pigpio-pwm} <com.pi4j.plugin.pigpio.provider.pwm.PiGpioPwmProviderImpl> {com.pi4j.plugin.pigpio.provider.pwm.PiGpioPwmProviderImpl} 
├─DIGITAL_OUTPUT: [1] <com.pi4j.io.gpio.digital.DigitalOutputProvider> 
│ └─PROVIDER: "PiGpio Digital Output (GPIO) Provider" {pigpio-digital-output} <com.pi4j.plugin.pigpio.provider.gpio.digital.PiGpioDigitalOutputProviderImpl> {com.pi4j.plugin.pigpio.provider.gpio.digital.PiGpioDigitalOutputProviderImpl} 
├─SERIAL: [1] <com.pi4j.io.serial.SerialProvider> 
│ └─PROVIDER: "PiGpio Serial Provider" {pigpio-serial} <com.pi4j.plugin.pigpio.provider.serial.PiGpioSerialProviderImpl> {com.pi4j.plugin.pigpio.provider.serial.PiGpioSerialProviderImpl} 
├─ANALOG_OUTPUT: [0] <com.pi4j.io.gpio.analog.AnalogOutputProvider> 
├─SPI: [1] <com.pi4j.io.spi.SpiProvider> 
│ └─PROVIDER: "PiGpio SPI Provider" {pigpio-spi} <com.pi4j.plugin.pigpio.provider.spi.PiGpioSpiProviderImpl> {com.pi4j.plugin.pigpio.provider.spi.PiGpioSpiProviderImpl} 
├─DIGITAL_INPUT: [1] <com.pi4j.io.gpio.digital.DigitalInputProvider> 
│ └─PROVIDER: "PiGpio Digital Input (GPIO) Provider" {pigpio-digital-input} <com.pi4j.plugin.pigpio.provider.gpio.digital.PiGpioDigitalInputProviderImpl> {com.pi4j.plugin.pigpio.provider.gpio.digital.PiGpioDigitalInputProviderImpl} 
└─I2C: [1] <com.pi4j.io.i2c.I2CProvider> 
  └─PROVIDER: "PiGpio I2C Provider" {pigpio-i2c} <com.pi4j.plugin.pigpio.provider.i2c.PiGpioI2CProviderImpl> {com.pi4j.plugin.pigpio.provider.i2c.PiGpioI2CProviderImpl} 
----------------------------------------------------------


Pi5
Pi4J PROVIDERS
----------------------------------------------------------
PROVIDERS: [6] "I/O Providers" <com.pi4j.provider.impl.DefaultProviders> 
├─ANALOG_OUTPUT: [0] <com.pi4j.io.gpio.analog.AnalogOutputProvider> 
├─DIGITAL_OUTPUT: [1] <com.pi4j.io.gpio.digital.DigitalOutputProvider> 
│ └─PROVIDER: "GpioD Digital Output (GPIO) Provider" {gpiod-digital-output} <com.pi4j.plugin.gpiod.provider.gpio.digital.GpioDDigitalOutputProviderImpl> {com.pi4j.plugin.gpiod.provider.gpio.digital.GpioDDigitalOutputProviderImpl} 
├─SPI: [1] <com.pi4j.io.spi.SpiProvider> 
│ └─PROVIDER: "PiGpio SPI Provider" {pigpio-spi} <com.pi4j.plugin.pigpio.provider.spi.PiGpioSpiProviderImpl> {com.pi4j.plugin.pigpio.provider.spi.PiGpioSpiProviderImpl} 
├─PWM: [1] <com.pi4j.io.pwm.PwmProvider> 
│ └─PROVIDER: "LinuxFS PWM Provider" {linuxfs-pwm} <com.pi4j.plugin.linuxfs.provider.pwm.LinuxFsPwmProviderImpl> {com.pi4j.plugin.linuxfs.provider.pwm.LinuxFsPwmProviderImpl} 
├─ANALOG_INPUT: [0] <com.pi4j.io.gpio.analog.AnalogInputProvider> 
├─DIGITAL_INPUT: [1] <com.pi4j.io.gpio.digital.DigitalInputProvider> 
│ └─PROVIDER: "GpioD Digital Input (GPIO) Provider" {gpiod-digital-input} <com.pi4j.plugin.gpiod.provider.gpio.digital.GpioDDigitalInputProviderImpl> {com.pi4j.plugin.gpiod.provider.gpio.digital.GpioDDigitalInputProviderImpl} 
├─SERIAL: [1] <com.pi4j.io.serial.SerialProvider> 
│ └─PROVIDER: "PiGpio Serial Provider" {pigpio-serial} <com.pi4j.plugin.pigpio.provider.serial.PiGpioSerialProviderImpl> {com.pi4j.plugin.pigpio.provider.serial.PiGpioSerialProviderImpl} 
└─I2C: [1] <com.pi4j.io.i2c.I2CProvider> 
  └─PROVIDER: "LinuxFS I2C Provider" {linuxfs-i2c} <com.pi4j.plugin.linuxfs.provider.i2c.LinuxFsI2CProviderImpl> {com.pi4j.plugin.linuxfs.provider.i2c.LinuxFsI2CProviderImpl} 
```

### CODE REVIEW

**Review by eitch — COMMENTED** [review]
You seem to be a C programmer by heart, using the `int rval = 0;` at the beginning of the method. I think this can be changed to a simple;

```
BoardModel model = BoardModelDetection.current().getBoardModel();
if (model.getSoc() != Soc.BCM2712) {
  return x;
} else {
  return y;
}
```

But my bigger question is: Should we really have a different priority for the pigpio provider? What would the benefit of this be? Currently we know users will have a better experience with  the Linux FS provider on a Raspberry Pi 5, as the Pigpio doesn't work. But if the user is on a Raspberry Pi 4, then it just depends which dependencies are added to their project.

We currently have three providers for PWM: raspberry, pigpio and linuxfs. The first is basically a stub, the latter the only that works on the Raspberry Pi 5. So in my opinion, that  would be a reason to change the priority based on the Board type.

@FDelporte what is your opinion?

**Comment by eitch** [comment]
Furthermore, the `model.getSoc() != Soc.BCM2712` is a bit cryptic, don't we have a convenience method isPi5()?

**Comment by FDelporte** [comment]
> Furthermore, the `model.getSoc() != Soc.BCM2712` is a bit cryptic, don't we have a convenience method isPi5()?

Good idea, I will add that to the BoardDetection

**Comment by FDelporte** [comment]
> But if the user is on a Raspberry Pi 4, then it just depends which dependencies are added to their project.

From the tickets it's clear that adding the right dependencies is a problem for a lot of users. So I think the extra priority decision in the plugin can help to tell users to "include all dependencies" and let Pi4J decide which will be used.

**Comment by taartspi** [comment]
The linuxfs pwm does work on pi4 but only HW and the address having a different meaning      I think the old code had linuxfs i2c higher than pigpio      I am not sure what is less trouble for users moving to this code base if they have all providers in their pom file. I thought this made sense but Robert u don't and Frank i am not sure your preference    I can agree with any plan.   So maybe the way to explain your thoughts is us the pi4 output to say what provider u want for each ioType

**Comment by taartspi** [comment]
@eitch @FDelporte Forgot to add your names so u got notified

**Comment by eitch** [comment]
Ok, then lets keep it like this

File: plugins/pi4j-plugin-gpiod/src/main/java/com/pi4j/plugin/gpiod/provider/gpio/digital/GpioDDigitalInputProviderImpl.java [hunk 1]
```diff
@@ -34,8 +36,14 @@ public DigitalInput create(DigitalInputConfig config) {
...
     @Override
     public int getPriority() {
-        // GpioD should be used if available
-        return 150;
+        // the gpioD driver should be higher priority when on Rp1 chip
+        int rval = 0;
+         if(BoardInfoHelper.usesRP1()) {
+            rval = 150;
+        }else{
+            rval = 150;
```
Comment (FDelporte) [inline plugins/pi4j-plugin-gpiod/src/main/java/com/pi4j/plugin/gpiod/provider/gpio/digital/GpioDDigitalInputProviderImpl.java:44]: value is the same in both if cases?

File: plugins/pi4j-plugin-gpiod/src/main/java/com/pi4j/plugin/gpiod/provider/gpio/digital/GpioDDigitalOutputProviderImpl.java [hunk 2]
```diff
@@ -67,8 +68,14 @@ public DigitalOutput create(DigitalOutputConfig config) {
 
     @Override
     public int getPriority() {
-        // GpioD should be used if available
-        return 150;
+        // the gpioD driver should be higher priority when on RP1 chip
+        int rval = 0;
+       if(BoardInfoHelper.usesRP1()) {
+            rval = 150;
```
Comment (FDelporte) [inline plugins/pi4j-plugin-gpiod/src/main/java/com/pi4j/plugin/gpiod/provider/gpio/digital/GpioDDigitalOutputProviderImpl.java:74]: idem

File: plugins/pi4j-plugin-linuxfs/src/main/java/com/pi4j/plugin/linuxfs/provider/i2c/LinuxFsI2CProviderImpl.java [hunk 3]
```diff
@@ -47,8 +48,14 @@ public LinuxFsI2CProviderImpl() {
 
     @Override
     public int getPriority() {
-        // the linux FS I2C driver should be used over the pigpio
-        return 150;
+       // the linux FS driver should be higher priority when on RP1 chip
+       int rval = 0;
+       if(BoardInfoHelper.usesRP1()) {
+            rval = 150;
```
Comment (FDelporte) [inline plugins/pi4j-plugin-linuxfs/src/main/java/com/pi4j/plugin/linuxfs/provider/i2c/LinuxFsI2CProviderImpl.java:54]: idem

File: plugins/pi4j-plugin-pigpio/src/main/java/com/pi4j/plugin/pigpio/provider/spi/PiGpioSpiProviderImpl.java [hunk 4]
```diff
@@ -56,8 +56,15 @@ public PiGpioSpiProviderImpl(PiGpio piGpio) {
 
     @Override
     public int getPriority() {
-        // the pigpio SPI driver should be used over the default
-        return 100;
+        // the Pigpio driver should be higher priority when NOT on RP1 chip.
+        int rval = 0;
```
Comment (FDelporte) [inline plugins/pi4j-plugin-pigpio/src/main/java/com/pi4j/plugin/pigpio/provider/spi/PiGpioSpiProviderImpl.java:60]: can be simplified as

```
        int rval = 100;
        if (BoardInfoHelper.usesRP1()) {
            rval = 50;
        }
```

File: plugins/pi4j-plugin-gpiod/src/main/java/com/pi4j/plugin/gpiod/provider/gpio/digital/GpioDDigitalInputProviderImpl.java (hunk 1 above)
Comment (taartspi) [reply plugins/pi4j-plugin-gpiod/src/main/java/com/pi4j/plugin/gpiod/provider/gpio/digital/GpioDDigitalInputProviderImpl.java:44]: Yes,  in both are the same to match what Robert did for 2.5     So I did the update and did pay attention to the fact both legs were equal.
Fixed

File: plugins/pi4j-plugin-gpiod/src/main/java/com/pi4j/plugin/gpiod/provider/gpio/digital/GpioDDigitalOutputProviderImpl.java (hunk 2 above)
Comment (taartspi) [reply plugins/pi4j-plugin-gpiod/src/main/java/com/pi4j/plugin/gpiod/provider/gpio/digital/GpioDDigitalOutputProviderImpl.java:74]: Should be the same as above.
fixed

**Pull request merged**